### ISSUE TITLE
Issue with adding validation series.

### ISSUE THREAD

**Issue by optionsraghu** [unit8co/darts#1494]
I am trying to add a validation series as mentioned in the documentation. I get the following error.

_ValueError: The dimensions of the series in the training set and the validation set do not match._

 I am not sure where I am going wrong. Kindly guide me. Many thanks. Here is the snippet of code:

from darts.dataprocessing.transformers import Scaler
from darts.utils.timeseries_generation import datetime_attribute_timeseries
#train val test split
train_series,valseries,test_series=series[0:388],series[388:480],series[480:]

train_scaler = Scaler()
scaled_train = train_scaler.fit_transform(train_series)
scaled_val=train_scaler.transform(valseries)                                                               
scaled_test = train_scaler.transform(test_series)

train_covs=scaled_train['cov1'].stack(scaled_train['cov2']).stack(scaled_train['cov3']).stack(scaled_train['cov4'])
val_covs = scaled_val['cov1'].stack(scaled_val['cov2']).stack(scaled_val['cov3']).stack(scaled_val['cov4'])
test_covs=scaled_test['cov1'].stack(scaled_test['cov2']).stack(scaled_test['cov3']).stack(scaled_test['cov4'])

model_pastcov = NBEATSModel(
    input_chunk_length=30,
    output_chunk_length=6,
    generic_architecture=True,
    num_stacks=50,
    num_blocks=2,
    num_layers=6,
    layer_widths=512,
    n_epochs=100,
    nr_epochs_val_period=1,
    batch_size=800,
    model_name="nbeats_run",
)

model_pastcov.fit(
    series=scaled_train['label'],
    past_covariates=[train_covs],
    val_series=scaled_val,
    verbose=True,
);

![image](https://user-images.githubusercontent.com/15772927/213159749-edf14ea6-f750-437e-8f57-ed0fa24b5520.png)

Basically I am trying to hold out a validation series that needs to be used when training goes on. Once again thanks for the help. Another query in the same lines is: Does Darts automatically take the past values of labels in its learning or do I need to stack it as an additional covariate ?

**Comment by madtoinou** [comment]
Hi,

It's an excellent habit to retain a part of the data for validation and test, the validation set can actually be leveraged by Darts to save the best checkpoint.

In your call to `fit`, it seems like you pass `scaled_train['label']` to the `series` argument whereas for the `val_series` argument, you pass `scaled_val`. The discrepancy in the dimensions probably comes from here.

I am not sure to understand what you mean by the "past values of labels" but everything contained in the `TimeSeries` provided as input will be used to train the model, you should not provide them as covariate.

On a side note, you could use `concatenate` ([documentation](https://unit8co.github.io/darts/generated_api/darts.timeseries.html#darts.timeseries.concatenate)) to simply your chain of to `stack` to a single function.

Let me know if changing this solve your problem.

**Comment by optionsraghu** [comment]
Hi @madtoinou , Thank you for your response. I think I have missed passing the val series in the earlier query. fit. What I have done below is to work out a toy example of air quality and I am pasting the code below. Please guide if this is the right way to fit a model. Basically the dependent variable is NO2 in air quality and there are 4 independent variables. I have created a train/val/test split . Here I assume when I fit the model Darts automatically takes care of checking the training against validation set to build the final model. 

I also assume (please correct me) that the past data of NO2 (the depedent) variable is also taken into the model building implicitly without we stacking it onto the covariates? 

Here is the code:

import time
import sys
import numbers
import time
import math
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import datetime as dt
from functools import reduce
import statsmodels.api as sm 
from scipy.stats import normaltest
from darts import TimeSeries
from darts.metrics import mape, mase, mae, mse, ope, r2_score, rmse, rmsle
from darts.utils.statistics import check_seasonality, plot_acf, plot_residuals_analysis
from darts.dataprocessing.transformers.boxcox import BoxCox
from darts.utils.likelihood_models import GaussianLikelihood
import warnings
warnings.filterwarnings("ignore")
import logging
logging.disable(logging.CRITICAL)


df=pd.read_csv("C:/Users/QuantCues/Documents/airq.csv")

df=df.dropna()

df=df[["CO(GT)",'C6H6(GT)','NO2(GT)','RH','AH']]

df.index = pd.RangeIndex(start=0, stop=len(df), step=1)


series=TimeSeries.from_dataframe(df)

from darts.dataprocessing.transformers import Scaler
from darts.utils.timeseries_generation import datetime_attribute_timeseries
#train val split
train_series,valseries,test_series=series[0:8500],series[8500:9000],series[9000:]

train_scaler = Scaler()
scaled_train = train_scaler.fit_transform(train_series)
scaled_val=train_scaler.transform(valseries)                                                               
scaled_test = train_scaler.transform(test_series)

train_covs=scaled_train['CO(GT)'].stack(scaled_train['C6H6(GT)']).stack(scaled_train['RH']).stack(scaled_train['AH'])
val_covs = scaled_val['CO(GT)'].stack(scaled_val['C6H6(GT)']).stack(scaled_val['RH']).stack(scaled_val['AH'])
test_covs=scaled_test['CO(GT)'].stack(scaled_test['C6H6(GT)']).stack(scaled_test['RH']).stack(scaled_test['AH'])

from darts.models import (
    RNNModel,
    NBEATSModel,
    BlockRNNModel,
    
)


model_pastcov = NBEATSModel(
    input_chunk_length=10,
    output_chunk_length=6,
    generic_architecture=True,
    num_stacks=5,
    num_blocks=2,
    num_layers=6,
    layer_widths=5,
    n_epochs=5,
    nr_epochs_val_period=1,
    batch_size=8,
    model_name="nbeats_run",
)

model_pastcov.fit(
    series=scaled_train['NO2(GT)'],
    past_covariates=[train_covs],
    val_series=scaled_val['NO2(GT)'],
    val_past_covariates = val_covs,
    verbose=True,
);

Many thanks for looking. Please also let me know if you need a hand in documentation. I am willing to help. Thank you for youyr kind perusal.

**Comment by madtoinou** [comment]
Hi,

Sorry for the delay, I missed the notification.

There is a small typo in the call to fit, the `past_covariates` value is between `[]` whereas the `val_past_covariates` isn't but other than that, you code looks great in my opinion.

Don't hesitate to have a look at this [notebook](https://unit8co.github.io/darts/quickstart/00-quickstart.html), where important methods such as `historical_forecasts()` are presented.
